### ISSUE TITLE
Move video playback out of core and into an officially supported GDExtension

### ISSUE THREAD

**Issue by Calinou** [godotengine/godot-proposals#3286]
*Note: This change was discussed with reduz and others and is probably good to implement.*

### Describe the project you are working on

The Godot editor :slightly_smiling_face: 

### Describe the problem or limitation you are having in your project

Video playback in Godot currently leaves a lot to be desired:

- It has a lot of bugs (see open `bug` issues for [Theora](https://github.com/godotengine/godot/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+theora+label%3Abug), [WebM](https://github.com/godotengine/godot/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+webm+label%3Abug)).
- There are performance issues due to not being able to benefit from hardware video decoding, especially on mobile hardware.
- It takes valuable space in the binary due to video decoding libraries being large. (libtheora and libvpx also have their own dependencies that could be removed from core, such as libopus.)
- It lacks features that are a deal-breaker in some projects such as seeking.
- The number of supported formats is fairly limited, with patent-encumbered formats not being supported.
- [Library updates are difficult for maintainers to perform.](https://github.com/godotengine/godot/issues/8334) In general, libvpx (used for VP8 and VP9 decoding) isn't exactly known to be easy to work with.
- It's [difficult to compile on unconventional platforms](https://github.com/godotengine/godot/issues/30914), making the build process unnecessarily complicated there.

We have very few contributors knowledgeable with video decoding libraries, so bug fixes and improvements are rarely seen nowadays.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

With GDExtension (the replacement of GDNative in 4.0), we can move video decoding to an officially supported add-on. This add-on will likely use FFmpeg like [godot-videodecoder](https://github.com/kidrigger/godot-videodecoder) currently does, but it may also use another library depending on code size, maintenance quality and licensing.

There are many benefits to moving video playback out of core:

- The binary size penalty is removed from the Godot editor and export template binaries. Instead, the size cost is moved to the add-on's compiled libraries.
- Hardware video decoding could be used on compatible hardware and software.
- Support for patent-encumbered formats can be exposed optionally. (If this is implemented, this will be disabled by default.)
  - Users will need to check their country's regulations and possibly acquire licenses, especially for commercial use. It *is* possible for open source software to support patent-encumbered formats – otherwise, VLC would not be able to play back `.mp4` videos :slightly_smiling_face: 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Perform a change like https://github.com/godotengine/godot/pull/52003, but for VideoPlayer.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No.

### Is there a reason why this should be core and not an add-on in the asset library?

Video decoding needs to have hooks in the engine to be efficiently implemented, so it needs dedicated GDExtension work.

**Comment by sairam4123** [comment]
Do this for `VehicleBody3D` too.

**Comment by Calinou** [comment]
> Do this for `VehicleBody3D` too.

This should be discussed in a separate proposal.

**Comment by fire** [comment]
I'll contact anyone still on the video decoder team at https://github.com/kidrigger/godot-videodecoder.

@kidrigger @jamie-pate

**Comment by jamie-pate** [comment]
For the video extension, i just noticed that one of my devs is running Apple silicon but we didn't notice the gdnative dylib failed to load because it falls back to the built in playback (graceful degradation since it doesn't support stuff like steam position)

We also have no solution currently for web or mobile.

To do this properly we probably also want yuv texture support and then we can use the hw accelerated decoders.

**Comment by Catchawink** [comment]
@kidrigger any updates on this or ways I could possibly help? I'm looking to address #2553 and it appears to be contingent on this one.

**Comment by kidrigger** [comment]
@Catchawink https://github.com/godotengine/godot/pull/62737
The PR needs to be merged.
The WebM plugin is working pretty well otherwise.

Resolving conflicts right now: ETA < 1 hr.

Who should I asked to review this quickly?

**Comment by hani09876** [comment]
Please make it able to play video from url . Like YouTube Facebook and Twitter

**Comment by akien-mga** [comment]
@hani09876 Please refrain from commenting on proposals with off-topic content. This proposal is about making video playback a GDExtension. Streaming videos from the web is completely out of scope for this proposal.

This is also not the first time we've had to remind you of not posting off-topic or low effort comments, so please pay attention and be more respectful of the time our contributors have to spend dealing with your comments and proposals. You seem very eager and have lots of ideas, but GitHub is our main work tool and off-topic comments or low quality proposals are a distraction which prevent contributors from actively improving the engine. It's best if you start by discussing your ideas or issues with the community at large on one of the [other community platforms](https://godotengine.org/community/) (e.g. Discord or Facebook), before asking the engine contributors to help you.

If you do have an idea for a new feature that you think should be discussed further on GitHub, then please open an open-ended discussion in the dedicated Discussions forum: https://github.com/godotengine/godot-proposals/discussions
Proposals are for well-researched ideas, usually with a technical implementation proposal. Your past proposals often had "no idea" as answer to required sections, which is a good hint that those should be free-form posts in the Discussions forum. See https://github.com/godotengine/godot-proposals#suggesting-improvements for details.

**Comment by DeeJayLSP** [comment]
Is the extension out? I think this proposal shouldn't have been closed before that.

**Comment by fire** [comment]
I'll reopen.

**Comment by kidrigger** [comment]
So, how does one go about making a plugin officially supported?

[This plugin has all the WebM code in working condition](https://github.com/kidrigger/Godot-video-reference)

**Comment by Calinou** [comment]
> So, how does one go about making a plugin officially supported?

We can transfer it to the @godotengine organization, as was done for https://github.com/godotengine/webrtc-native (for example).

**Comment by makemefeelgr8** [comment]
> It lacks features that are a deal-breaker in some projects such as seeking.

Yup, it just killed one more project. I wonder what is the actual "kill count" in 2 years the issue have been open.
So, instead of chasing the impossible and unimplementable original approach, I suggest the easier way to get things done. At least for desktop PC platforms.

OpenCV.

Workflow would be like:
1. Use OpenCV to decode some frames.
2. Draw those frames on a texture.

Just how cool is that! Any video format is supported.
It will even deal with seeking, rewinds and so on.
And the effort to implement it is like non-existant. Compile the library with some flags and you're done.

**Comment by Calinou** [comment]
> OpenCV.

As mentioned above, large libraries like FFmpeg and OpenCV will not be considered for inclusion in core due to their binary size. What an extension uses is less important, but if you're able to integrate FFmpeg, integrating OpenCV isn't really worth the effort.

**Comment by makemefeelgr8** [comment]
It turned out to be way easier than expected. Here's the OpenCV PoC:

```c#
using Godot;
using OpenCvSharp; // Install https://www.nuget.org/packages/OpenCvSharp4.Windows
using System;

public partial class VideoSprite : Sprite2D
{
    [Export]
    public string VideoPath { get; set; }
    private VideoCapture _capture;
    private double _playbackPosition;

    public override void _Ready()
    {
        _capture = new VideoCapture(this.VideoPath);
        if (!_capture.IsOpened())
            throw new Exception($"Failed to open {this.VideoPath}");
        _playbackPosition = 0;
    }

    public override void _Process(double delta)
    {
        _playbackPosition += delta;
        var currentFrame = (uint)(_playbackPosition * _capture.Fps);
        if (_capture.PosFrames >= currentFrame)
            return; // It's already displaying the correct frame

        var frame = new Mat();
        _capture.Read(frame);
        if (frame.Empty())
            return; // Video is over

        // Convert frame data to godot boilerplate
        var pngBytes = frame.ToBytes(ext: ".png");
        var boilerplate = new Godot.Image();
        boilerplate.LoadPngFromBuffer(pngBytes);
        var texture = ImageTexture.CreateFromImage(boilerplate);
        this.Texture = texture;
    }
}
```

It results in about 15fps now, because of the stupid conversions, that go like:

Video Frame -> Raw OpenCV -> OpenCV mat -> png bytes -> godot image -> godot texture -> Sprite2D.

If one were to shave a few steps from the list, it would yield an acceptable performance for sure.
Also, one would have to decouple video frame processing from the rendering thread (for obvious reasons).

**Comment by jamie-pate** [comment]
To get decent performance (hw acceleration) you will want to write yuv
video frames directly to the texture using opengl or vulkan and then use a
shader to convert to rgb. Last time I looked there was no path to do that
in godot..

Any other path for decoding video from standard formats will incur a lot of
overhead.

How much overhead is acceptable depends on the size of your video frames vs
the size of your cpu...

**Comment by makemefeelgr8** [comment]
> To get decent performance (hw acceleration) you will want to write yuv video frames directly to the texture using opengl or vulkan and then use a shader to convert to rgb. Last time I looked there was no path to do that in godot.. Any other path for decoding video from standard formats will incur a lot of overhead. How much overhead is acceptable depends on the size of your video frames vs the size of your cpu...

There's an easier way. I just wish godot exposed some raw opengl functions. I could pass the pointer to OpenCV output, bypassing the whole conversion flow.

```cpp
glTexImage2D(
  GL_TEXTURE_2D,
  0,
  GL_RGB,
  image.cols,
  image.rows,
  0,
  GL_BGR,
  GL_UNSIGNED_BYTE,
  image.ptr()
);
```

> Video Frame -> Raw OpenCV -> OpenCV mat -> bmp bytes -> godot image -> godot texture -> Sprite2D -> OpenGL

would turn into

> Video Frame -> Raw OpenCV -> OpenGL

But (unlike Unity where GL is exposed) Godot faced no AAA gamedev, so, I assume there was no need for opengl functions in c#. Or for video playback.

**Comment by fire** [comment]
There's an external texture code, @BastiaanOlij but I don't know if it's usable for this purpose.

**Comment by BastiaanOlij** [comment]
Is this Godot 3 or 4? 

I don't have any experience with C#, only C++/GDExtension when it comes to this. Seeing how OpenGL works, you can just include OpenGL and you should be running within the correct context. There may be some threading issues and we do plan on improving how OpenGL is exposed in Godot 4.

In Godot 3 you can use `VisualServer.texture_get_texid` to retrieve the OpenGL texture ID for an existing texture, then just use OpenGL commands to load in data.
In Godot 4 you can use `RenderingServer.texture_get_native_handle` to do the same.

In both cases you would use the VisualServer/RenderingServer to create a new texture object, and then load data in. 

There was more logic I was working on for ARCore but that never got merged as we never managed to get ARCore working as a plugin.

**Comment by DeeJayLSP** [comment]
The only advantage I can see in OpenCV over FFmpeg might be licensing (permissive over copyleft).

However, unless you plan to export to consoles (or any platform with NDAs) I don't see the issue with FFmpeg as its license allows not disclosing source in case of shared libs.

*Now somewhat breaking the topic*

When it comes to multiplatform exporting that includes consoles (most of them have some built-in media libraries you can link against) I'd say a custom wrapper is a better option than any of the above.

**Comment by makemefeelgr8** [comment]
> RenderingServer.texture_get_native_handle

I'm talking Godot 4. Thanks for trying to help. I tried using the `RenderingServer` and `RenderingDevice` combo, except none of it worked. For example, this one always returns an empty array:

```c#
var txData = RenderingServer.GetRenderingDevice().TextureGetData(_texture.GetRid(), 0);
```

This one always throws:
```c#
RenderingServer.GetRenderingDevice().TextureUpdate(_texture.GetRid(), 0, _rawData);
```

And method description is as cryptic as they make them:

```C#
        // Summary:
        //     Returns the texture data for the specified layer as raw binary data. For 2D textures
        //     (which only have one layer), layer must be 0.
        //     Note: texture can't be retrieved while a draw list that uses it as part of a
        //     framebuffer is being created. Ensure the draw list is finalized (and that the
        //     color/depth texture using it is not set to Godot.RenderingDevice.FinalAction.Continue)
        //     to retrieve this texture. Otherwise, an error is printed and a empty System.Byte[]
        //     is returned.
        //     Note: texture requires the Godot.RenderingDevice.TextureUsageBits.CanCopyFromBit
        //     to be retrieved. Otherwise, an error is printed and a empty System.Byte[] is
        //     returned.
```

It's an anti-pattern. `Ensure the draw list is finalized`. How? `texture requires the Godot.RenderingDevice.TextureUsageBits.CanCopyFromBit to be retrieved`. How?

It should be like `Ensure the draw list is finalized by calling X.EnsureFinalized()`.

I also tried switching the engine to opengl, and writing some texture bytes using `OpenTK`, like:

```C#
        _textureHandle = RenderingServer.TextureGetNativeHandle(_texture.GetRid());
        GL.BindTexture(TextureTarget.Texture2D, (int)_textureHandle);
        GL.TexSubImage2D(TextureTarget.Texture2D, 0, 0, 0, frame.Cols, frame.Rows, PixelFormat.Bgra, PixelType.UnsignedByte, _rawData);
        GL.BindTexture(TextureTarget.Texture2D, 0);
```

But it just threw some access violations & memory errors from the unmanaged part. The context might be the reason, though, as I init the library like:
```c#
        GL.LoadBindings(new OpenTK.Windowing.GraphicsLibraryFramework.GLFWBindingsContext());
```

I assume there's a need to grab GL context from godot somehow, for it to actually draw something.

**Comment by makemefeelgr8** [comment]
> The only advantage I can see in OpenCV over FFmpeg might be licensing (permissive over copyleft).

With OpenCV you're also getting a bunch of powerful tools for image editing, object recognition, machine learning, and so on.

**Comment by DeeJayLSP** [comment]
> With OpenCV you're also getting a bunch of powerful tools for image editing, object recognition, machine learning, and so on.

Yes, but the discussion here is mainly about video decoding. Currently there's no huge demand for those features. And if someone really needs it, they can always extend the engine (for video decoding as Calinou mentioned, there's already a [FFmpeg-based video decoder by EIRTeam](https://github.com/EIRTeam/EIRTeam.FFmpeg)).

Also, to correct my previous statement, I peeked into OpenCV's source code and it seems to make use of FFmpeg itself for video decoding, therefore the copyleft licensing part also applies to OpenCV. If one wants to do video decoding with it, why not just use FFmpeg directly?

**Comment by makemefeelgr8** [comment]
> there's already a [FFmpeg-based video decoder by EIRTeam](https://github.com/EIRTeam/EIRTeam.FFmpeg)

But the implementation is horrible! It's no different from my code above. It's even worse. Have you seen their code? Those guys copy a video frame [line by line in a loop](https://github.com/EIRTeam/EIRTeam.FFmpeg/blob/master/video_decoder.cpp#L456C6-L456C6). And then they use [image.set_data](https://github.com/EIRTeam/EIRTeam.FFmpeg/blob/master/video_decoder.cpp#L465) and [image texture. update](https://github.com/EIRTeam/EIRTeam.FFmpeg/blob/master/video_decoder.cpp#L483). They lock a mutex 3 times per frame.

**Comment by Morpheu5** [comment]
> Yup, it just killed one more project. I wonder what is the actual "kill count" in 2 years the issue have been open.

Sorry to add to the pile but it did just kill [another one](https://forum.godotengine.org/t/improve-performance-of-spotlight-based-fake-gi-on-mobile/44300?u=morpheu5). It's a shame but I'm happy to help if I can with testing and such if necessary.

**Comment by JanWerder** [comment]
Wouldn't including libvlc be an alternate solution? (LGPL2.1 license)
There's one project that has done that: https://github.com/RancidMilkGames/Godot-VLC

**Comment by Calinou** [comment]
> Wouldn't including libvlc be an alternate solution? (LGPL2.1 license) There's one project that has done that: [RancidMilkGames/Godot-VLC](https://github.com/RancidMilkGames/Godot-VLC)

LGPL libraries can't be integrated in core: https://docs.godotengine.org/en/stable/contributing/development/best_practices_for_engine_contributors.html#don-t-use-complex-canned-solutions-for-simple-problems

> Libraries must use a permissive enough license to be included into Godot. Some examples of acceptable licenses are Apache 2.0, BSD, MIT, ISC, and MPL 2.0. In particular, we cannot accept libraries licensed under the GPL or LGPL since these licenses effectively disallow static linking in proprietary software (which Godot is distributed as in most exported projects). This requirement also applies to the editor, since we may want to run it on iOS in the long term. Since iOS doesn't support dynamic linking, static linking is the only option on that platform.

While they could be used in an official extension (as extensions are dynamically linked), dynamic linking isn't allowed on all platforms. Also, I think Godot should remain fully permissively licensed, including official extensions to avoid licensing pitfalls that other projects have encountered (such as Qt).

**Comment by lostminds** [comment]
As finding a decoder library with suitable features, license and platform support seems difficult, could an alternative be to  instead rely on platform specific built in system services for this? At least for MP4 / H264 there seems to be support for this in Windows via [Microsoft Media Foundation](https://learn.microsoft.com/en-us/windows/win32/medfound/mpeg-4-support-in-media-foundation), macOS/iOS via [CoreVideo](https://developer.apple.com/documentation/corevideo?language=objc), Android via [MediaCodec](https://developer.android.com/media/platform/supported-formats) and possibly in some linux versions as well?

The tradeoff would of course be that more platform specific code would be needed, and feature support will vary. And for web deployments for example this might not be possible. But even a lowest common denominator feature set will likely be a big improvement over the current support, and while not ideal I'm sure it won't be the only feature that is not available on all platforms. One issue though is that the currently only supported format OGV/Theora is likely not supported by any of these system services, since it's so unusual. So moving exclusively to using system decoders would break compatibility with existing projects using these videos.

**Comment by voylin** [comment]
@Calinou I have successfully made a GDExtension which has video and audio playback with seeking. Not certain if this could be turned into an official extension though. I know a lot more work would be necessary to make it more user friendly. It uses FFmpeg 

If I were to be allowed to work on an official gdextension, how would the process be? I'd probably also look into #8049 to see if it's possible for me to also implement that as I both need it for my video editor and as it could improve performance for people using the GDExtension.

**Comment by voylin** [comment]
A little update on this, I got a GDExtension working for creating video playback. It is not a drop in replacement as you still need to write the code yourself for displaying the frames in order. But it is a beginning. https://github.com/VoylinsGamedevJourney/gde_gozen

Only question would be is if I can turn this into an officially supported GDExtension as I don't know how it can become "officially supported" ^^" So if you think I'm on the right track, let me know and I can try to create a Player node ;)

**Comment by Calinou** [comment]
> Only question would be is if I can turn this into an officially supported GDExtension

I think the extension would need to be production-ready first, which means it should be able to work as a drop-in replacement for the existing video playback system (other than reencoding your videos to the desired format).

**Comment by voylin** [comment]
> I think the extension would need to be production-ready first, which means it should be able to work as a drop-in replacement for the existing built-in video playback system (other than reencoding your videos to the desired format).

If it were to use FFmpeg as a library, is this okay for an official GDExtension?

**Comment by kobligo** [comment]
From [https://www.ffmpeg.org/legal.html](https://www.ffmpeg.org/legal.html)
> FFmpeg is licensed under the [GNU Lesser General Public License (LGPL) version 2.1](http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html) or later. However, FFmpeg incorporates several optional parts and optimizations that are covered by the [GNU General Public License (GPL) version 2](http://www.gnu.org/licenses/old-licenses/gpl-2.0.html) or later. If those parts get used the GPL applies to all of FFmpeg.

I know nothing about licenses. Is this compatible or not?
Isn't FFMPEG Godot's only real option when it comes to creating a proper video player?

**Comment by DeeJayLSP** [comment]
> I know nothing about licenses. Is this compatible or not? Isn't FFMPEG Godot's only real option when it comes to creating a proper video player?

In order to play a video, you have to:
1. Demux the container
2. Decode the video stream with the correct decoder
3. Same as above but for the audio stream

FFmpeg is a fairly common implementation that mostly makes the 3 things above easy. A video player could be done by directly implementing the decoders and demuxers into a VideoStream class, but doing so would require a bit more work.

**Comment by fire** [comment]
I would prefer to keep one video codec in the core of godot engine for testing purposes.

**Comment by Calinou** [comment]
> I know nothing about licenses. Is this compatible or not?

It is compatible with Godot's license, but LGPL imposes restrictions on dynamic linking. This is why it can't be integrated in core and should be an add-on instead, as not everyone can deal with these restrictions (e.g. iOS exports due to static linking being mostly nonexistent there).

**Comment by jamie-pate** [comment]
> I would prefer to keep one video codec in the core of godot engine for testing purposes.

I agree that the simplest, most compatible (platform/license wise, not content wise) codec is a good idea for broad compatibility and a reference implementation. If a developer really really needs video to work everywhere they can re-encode it to fit the engine/performance constraints! Re-encoding is a much easier solution than broad support.

It would be nice to have platform integration when available though, for devs who are targeting specific platforms:

Vulkan now has an extension for video decoding!
https://github.com/KhronosGroup/Vulkan-Docs/blob/main/proposals/VK_KHR_video_queue.adoc#3-proposal
This seems like a great direction to shoot for long term.
Design-wise (for our api) the features of their extension would probably be well thought out and offer maximum performance + utility.

Windows/Linux drivers for AMD/NVIDIA/Intel all support the extension
https://github.com/mpv-player/mpv/discussions/13909 has more research on desktop compatibility (as of May 29, 2023) (maybe mpv-player can be used as an integration example)

Mac/iOS/Android would have to have an alternate implementation. Preferably one that can hardware decode directly to video memory.

[Android 15:](https://source.android.com/docs/compatibility/15/android-15-cdd#1_introduction)

> MUST NOT enumerate support for the VK_KHR_video_queue, VK_KHR_video_decode_queue, or VK_KHR_video_encode_queue extensions.
(probably because they want you to use https://developer.android.com/reference/android/graphics/SurfaceTexture or [VK_ANDROID_external_memory_android_hardware_buffer](https://registry.khronos.org/vulkan/specs/latest/html/vkspec.html#VK_ANDROID_external_memory_android_hardware_buffer): https://stackoverflow.com/a/43550804/193232)

Metal seems like it has [similar capability](https://stackoverflow.com/a/58900720/193232)

https://www.khronos.org/blog/khronos-finalizes-vulkan-video-extensions-for-accelerated-h.264-and-h.265-encode

**Comment by DeeJayLSP** [comment]
I wonder if MPEG-1 would be a good choice. Compared to Theora, there is [a very simple and lightweight library for it](https://github.com/phoboslab/pl_mpeg/), decoding is significantly faster in single thread, and quality is just *slightly* worse.

Out of curiosity, I went to experiment with a few encoding parameters and resulted in me writing [this article](https://gamedev.djlsp.xyz/blog/mpeg-1-for-games/) with the results.

**Comment by jamie-pate** [comment]
My experiments led to the conclusion that vp9 software decoding is efficient enough for desktop use.

It can produce reasonable quality at very low bitrates  most of the time is actually spent converting the decoded pixels from yuv to rgb anyways so the video resolution plays a major role.

**Comment by Calinou** [comment]
Remember that we had VP8/VP9 decode in 3.x and had to remove it because it was buggy in practice. simplewebm/libvpx are much more complex to integrate than libogg/libtheora and they require you to use their custom assembler (yasm) to achieve good performance.

**Comment by DeeJayLSP** [comment]
> simplewebm/libvpx are much more complex to integrate than libogg/libtheora

[Theora isn't free of bugs either](https://github.com/godotengine/godot/issues/66331).

In fact, it seems libtheora itself is quite buggy.  I recall someone converting a Theora video to WebM in a certain Godot issue, and it seems the bug carried over. I even have some problems playing a video on the mpv player.

Still in defense of MPEG-1, I have encoded a reference video into Theora and MPEG-1 using the best encoding settings I've spent hours trying to achieve, while trying to maintain a close size to the original (encoded in 1440p then downscaled):

https://www.youtube.com/watch?v=hBK_2V3Or70

(I swear I didn't do anything to make Theora look bad)

Additionally, like I mentioned, there is [pl_mpeg](https://github.com/phoboslab/pl_mpeg), which is a single header MPEG-1 implementation with SDL and GLEW examples, and I bet it would be much easier to maintain than Theora.

**Comment by jamie-pate** [comment]
It sounds like `libvpx` is pie in the sky since it was already removed for the same reasons.

I re-reviewed the core criteria of this proposal, and if the choice is:
a. removing video completely
b. switching to MPEG-1

Then I'd agree with MPEG-1 here. A `single header MPEG-1 implementation` with an MIT license _sounds_ like a slam dunk to solve these points without making things worse. Remember that when theora was added MPEG-1 would have still been license/patent encumbered. It's been hard for me to shed that bias!

How many of these still apply to `libtheora` vs `libvpx` (since libvpx is already gone)

> * It takes valuable space in the binary due to video decoding libraries being large. (libtheora and libvpx also have their own dependencies that could be removed from core, such as libopus.)
> * It lacks features that are a deal-breaker in some projects such as seeking.
> * https://github.com/godotengine/godot/issues/8334 In general, libvpx (used for VP8 and VP9 decoding) isn't exactly known to be easy to work with.
> * It's https://github.com/godotengine/godot/issues/30914, making the build process unnecessarily complicated there.

It also meets @fire 's request:

> [I would prefer](https://github.com/godotengine/godot-proposals/issues/3286#issuecomment-2585166699) to keep one video codec in the core of godot engine for testing purposes.

An officially supported extension for platform-level decoding of advanced formats would support the other original criteria:

> * The number of supported formats is fairly limited, with patent-encumbered formats not being supported.
> * There are performance issues due to not being able to benefit from hardware video decoding, especially on mobile hardware.

> But since I was talking about decoding speed

Looks like MPEG-1 is 4x faster, but VP9 gives 5x bitrate reduction with better quality. IMO that still a win for VP9! Unfortunately this is irrelevant if VP9 is off the table :D

A large portion of video playback frame time in godot is the colorspace conversion. Any benchmarks need to be run with the current `VideoStream` workflow to include that number. (increasing quadratically with resolution)

I'm really really curious how much the colorspace conversion would cost vs the decoding speed of MPEG-1

> March 2, 2022: I profiled software decoding and the cpu was **dominated** by yuv->rgb  (* on the main game thread) because in the ffmpeg plugin the decoding happens in a separate thread

You maybe can estimate this cost by adding an ffmpeg filter to convert the frames to RGB before sending them to /dev/null with `-threads 1`?

**Comment by fire** [comment]
I think the best we can do is use mpeg-1 and replace the code for libtheora with stubs that do nothing.

There's a rgba to yuva shader done by @EIREXE

https://github.com/EIRTeam/EIRTeam.FFmpeg/commit/49efa2a0ac032f5991e7514e0e73899eb18251f4

**Comment by Calinou** [comment]
> I think the best we can do is use mpeg-1 and replace the code for libtheora with stubs that do nothing.

I'd say removing Theora support in 4.x is out of the question, as it would not only break compatibility, but also require users to convert their existing videos to another lossy format (and therefore suffer from double compression).

**Comment by jamie-pate** [comment]
> I'd say removing Theora support in 4.x without providing it as an add-on is out of the question
Deprecate it for now? I don't think anyone's lining up to push `VideoStream` forward at the moment, so adding MPEG-1 would probably wait until someone's ready to do that work. It would probably be a good way to learn the systems involved also.

Personaly I've never had good experiences with Theora encoding and it's left a bad taste in my mouth when I've used it professionally. Neither MPEG-1 or Theora are going to be good for anything over 720p, but with proper docs we can steer users to accept the limitations.

> There's a rgba to yuva shader done by @EIREXE

(this is part of the 'push forward' VideoStream needs, YUV support. I'm sure there's other features that are missing though)

**Comment by DeeJayLSP** [comment]
If anyone is interested, I made [an attempt at implementing MPEG-1 support through pl_mpeg](https://github.com/DeeJayLSP/godot/tree/plm). Almost nothing works, but at least you can set a .mpg file into VideoStreamPlayer without crashing.

The only thing that is working 100% is probably demuxing (it detects width*height correctly).

**Comment by lostminds** [comment]
Having struggled a bit with using the currently implemented video features for a number of demo/tutorial video snippets in a project I can perhaps offer a "user" perspective on this. I would support the original idea of moving to something like ffmpeg (or similar), even if this due to legal reasons requires it to be a separate GDExtension. And with that just leave theora support as is for compatibility or remove it with deprecations warnings if the new external solution could also handle decoding and displaying theora.

The main issue as I see it with the current implementation is simply that the supported ogv/theora format is very unusual, which requires extra specific tools and steps in the development workflow to re-encode videos. To make the workflow less cumbersome and reduce the initial threshold to use these features I'd say the most important change would be to directly support common modern formats, at least for import (like mp4 containers with H.264 encoding).

The second issue I've had with the current implementation is the very limited display and interaction support, for example you need to start playing a video to get something to show in a VideoStreamPlayer (see https://github.com/godotengine/godot/issues/92050) and there's no support for searching. This severely limits how you can use videos in UIs and make them hard to work with in the editor. These issues are I think not directly dependent on encoding, but would likely be easier to fix with some external library like ffmpeg.

Third, I would also consider future possibilities of not only decoding/displaying video but also encoding video that integrating something like ffmpeg could offer. While of course displaying video will be more common, encoding video to modern formats would bring new opportunities. Both for editor/development workflows to render video assets and video "filter" style features in Godot projects using the rendering engine combined with importing and/or exporting common video formats.

**Comment by jamie-pate** [comment]
> I would support the original idea of moving to something like ffmpeg (or similar), even if this due to legal reasons requires it to be a separate GDExtension.

This exists (only has builds for windows64/linux64)
https://github.com/EIRTeam/EIRTeam.FFmpeg/tree/master

**Comment by fire** [comment]
@DeeJayLSP 

```
Error: .\thirdparty/misc/pl_mpeg.h(1390): error C2220: the following warning is treated as an error
Warning: .\thirdparty/misc/pl_mpeg.h(1390): warning C4996: 'fopen': This function or variable may be unsafe. Consider using fopen_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.
```

Will this require much code tinkering?

**Comment by DeeJayLSP** [comment]
> Will this require much code tinkering?

Would be much easier if there was a `*_NO_STDIO` define like most single-header libraries. That function isn't even used by that implementation, so we could just patch it out.

**Comment by fire** [comment]
I have my attempt at porting to 4.0 wmf. https://github.com/V-Sekai/godot/tree/archived/wmf

Which is based on Godot Engine 3. https://github.com/V-Sekai/godot/tree/vsk-wmf-3.5

**Comment by DeeJayLSP** [comment]
My `pl_mpeg` implementation is almost done. Playback seems to support both video and audio correctly.

There are two problems, however:
- `pl_mpeg`'s built-in YUV to RGB conversion seems extremely slow. I couldn't figure out a way to link it to yuv2rgb. A 720p video at 24FPS seems to work fine, but a 1440p 60FPS one will stutter a lot.
- Data is read from memory instead of file.

**Comment by fire** [comment]
Did you investigate https://github.com/EIRTeam/EIRTeam.FFmpeg/commit/49efa2a0ac032f5991e7514e0e73899eb18251f4?

**Comment by DeeJayLSP** [comment]
> Did you investigate [EIRTeam/EIRTeam.FFmpeg@49efa2a](https://github.com/EIRTeam/EIRTeam.FFmpeg/commit/49efa2a0ac032f5991e7514e0e73899eb18251f4)?

I was focusing on getting things working at an acceptable way. Using a shader would complicate things more (I have no idea of how to implement it).

**Comment by berarma** [comment]
I think there's value in having a well supported and highly compatible codec in the core. I've created godotengine/godot#101958 to address the issues with Theora video streams.

**Comment by DeeJayLSP** [comment]
> I think there's value in having a well supported and highly compatible codec in the core. I've created [godotengine/godot#101958](https://github.com/godotengine/godot/pull/101958) to address the issues with Theora video streams.

Guess I was wrong about libtheora being broken.

Oh well, I had a good time making that pl_mpeg implementation. Maybe I'll get it working on the best possible state and release as a custom module.

**Comment by berarma** [comment]
> > I think there's value in having a well supported and highly compatible codec in the core. I've created [godotengine/godot#101958](https://github.com/godotengine/godot/pull/101958) to address the issues with Theora video streams.
> 
> Guess I was wrong about libtheora being broken. It was the majority of implementations, including Godot's.
> 
> Oh well, I had a good time making that pl_mpeg implementation. Maybe I'll get it working at the best possible state and release as a custom module.

Sorry. I didn't know you were working on a replacement. I didn't imagine I would go that far when I started looking into this, I just wanted to take a glance at the code and so I didn't pay much attention around. It spiraled fast into "maybe there's a bug" and "maybe I can fix it".

Anyway, having other options for video codecs is very interesting and needed. I like the idea of being able to add new codecs via extensions.

**Comment by DeeJayLSP** [comment]
Module done: https://github.com/DeeJayLSP/godot-mpg

We can all go back to the subject of this proposal now.

> Sorry. I didn't know you were working on a replacement.

My excuse for suggesting to replace Theora with MPEG-1 was mostly because of the bugs. While MPEG-1 performs more than twice as fast, keeping Theora is better to maintain compatibility.

Could still work as a replacement for users who want to trade compression ratios for decoding speed.

**Comment by Joy-less** [comment]
A comprehensive list of options for video decoding:
- Theora: Poor compression, not supported in browsers
- H264 (MP4): Patented until 2030 (Brazil)
- H265 (MP4): Patented until 2030+
- AV1: Too slow to use in software decoding, not enough support in hardware decoding
- MPEG-1 (MPG): Worse compression than Theora, not supported in browsers
- MPEG-2 (MPG): Worse compression than Theora, not supported in browsers, patented in Malaysia until 2035
- VP8 (WEBM): Removed from Godot for being too difficult to integrate
- VP9 (WEBM): Removed from Godot for being too difficult to integrate
- FFmpeg: Licensing issues (GPL/LGPL), large binary size (80MB+)
- Operating System codecs: Different API for every OS, relies on OS libraries (which may change or be faulty)

In my opinion, the only realistic option in the near future for modern video decoding support is WEBM (VP8 and VP9). It's supported by every browser, has a permissive license, has high compression quality, has hardware acceleration, and is commonly used as an alternative to MP4 by non-tech users. While some of these may not apply to Godot (e.g. hardware acceleration) they are a reason why users may already have a WEBM file they wish to import into Godot.

> In general, libvpx (used for VP8 and VP9 decoding) isn't exactly known to be easy to work with.

I believe that many of the issues with WEBM were caused by architecture-specific optimisations (e.g. x86, ARM) which are complicated to resolve. For example, the pull request linked in the original post (https://github.com/godotengine/godot/issues/30914) is related to ARM-specific optimisations. In that case, the optimisations could be skipped, similar to https://github.com/godotengine/godot/pull/117921, and then potentially added later (perhaps only for certain platforms) if a contributor can figure it out.

Notably, many fixes & improvements to Theora support were added after this proposal. The number of contributors to Godot has increased since Unity's runtime fee drama in 2023. This could lead to architecture-specific optimisations and hardware acceleration being added later down the line by contributors.

> With GDExtension (the replacement of GDNative in 4.0), we can move video decoding to an officially supported add-on. This add-on will likely use FFmpeg like godot-videodecoder currently does, but it may also use another library depending on code size, maintenance quality and licensing.

I'm personally against the idea of an external addon that relies on FFmpeg for two reasons:
1. FFmpeg is a large library (80MB+) with a viral LGPL/GPL license. For games that only include one or two video files, the library size is too big to be worth it (in comparison, dedicated WEBM support could be less than 1MB). Additionally, the LGPL license is poorly understood and requires that the library is dynamically linked.
2. If video decoding is buggy when included in core, it will be more buggy as an extension. Most users wouldn't use the extension, so they wouldn't report any issues. I anticipate the extension fading away into a "half-abandoned" state.

Support for WEBM could be included in core with a minimal footprint:
- Remove architecture-specific optimisations, keeping only pure C code
- Remove encoding support, keeping only decoding support
- Remove VP8 support, keeping only VP9 support (optional)

**Comment by Calinou** [comment]
[s&box recently replaced FFmpeg usage with custom video codec implementation](https://sbox.game/news/update-26-04-22#media-framework-rewrite) (plus Media Foundation for H.264 through the system-provided decoder, but only on Windows). The commit in question: https://github.com/Facepunch/sbox-public/commit/2404e8aab65bb47c19f0ea07d9625377e6606043

> - AV1: Too slow to use in software decoding, not enough support in hardware decoding

Hardware AV1 decoding is starting to be fairly widespread (more so than encoding). The problem is that using it from a game engine is significantly more difficult than in a dedicated video player. Vulkan Video isn't very well-supported by drivers, especially outside of Linux.

Software AV1 decoding is also a lot more optimized these days, and CPUs have gotten faster compared to the codec's introduction in 2020. There's a good chance your CPU can decode 1080p60 or 1440p30 AV1 without dropping any frames. (From what I've seen, most games still ship video in 1080p these days, especially indie games.)

> - Remove architecture-specific optimisations, keeping only pure C code

Last time I checked, pure C code (i.e. no assembly) is really slow for VP9 decoding. It's not as bad as AV1 in pure C, but it probably makes VP9 decoding roughly as demanding as AV1 + assembly.

**Comment by Joy-less** [comment]
@Calinou I agree that if not VP8/VP9 (WEBM), then AV1 is the next-best option. Although, I still think VP8/VP9 would make more sense at this time:

> Hardware AV1 decoding is starting to be fairly widespread

I would disagree with this point. The earliest NVIDIA GPU with AV1 decoding support (not encoding) is the RTX 30-series (e.g. RTX 3050), which is still a relatively new GPU (2020). For reference, I have a GTX 1080, which can still run most AAA games and I don't have any intention to upgrade it soon.

> There's a good chance your CPU can decode 1080p60 or 1440p30 AV1 without dropping any frames.

My CPU (pretty old) can easily decode Theora at 1440p60 in Godot. I'm not sure about WEBM though. I would say 1080p60 is the minimum for most videos nowadays.

> Last time I checked, pure C code (i.e. no assembly) is really slow for VP9 decoding. It's not as bad as AV1 in pure C, but it probably makes VP9 decoding roughly as demanding as AV1 + assembly.

May be true, although I couldn't find any benchmarks for this. Libvpx has had a lot of improvements as well since it was removed from Godot. I wonder if there are some benchmarks somewhere?

Another thing to consider is that, unlike AV1, VP8/VP9 was already added to Godot before, so it would be a lot easier to re-implement it. Do the same concerns about difficulty to build assembly in VP8/VP9 not apply to AV1? Unlike AV1, VP8/VP9 is usable in pure C which would be a good template for future contributors to improve upon.

As far as I'm aware, there's not a standard container for AV1, so which container would Godot support (MP4, WEBM, MKV)? If WEBM, then wouldn't it make sense to support VP8/VP9 as well?

**Comment by Calinou** [comment]
> As far as I'm aware, there's not a standard container for AV1, so which container would Godot support (MP4, WEBM, MKV)?

Probably WebM, given it's what AV1 videos on the web typically use.

**Comment by Joy-less** [comment]
@Calinou Could I ask you what the team's thoughts are about this topic? I would be interested to at least try to re-add VP8/VP9 support, if it would be accepted.
On that topic, is the Opus pull request likely to be accepted?

**Comment by Calinou** [comment]
> Could I ask you what the team's thoughts are about this topic? I would be interested to at least try to re-add VP8/VP9 support, if it would be accepted.
> On that topic, is the Opus pull request likely to be accepted?

I'm personally in favor of both, but I can't guarantee either PR is merged. I'll ask around for opinions on the topic.
